### PR TITLE
Add br and cond_br operations to std

### PR DESCRIPTION
**Pull request by tobiasgrosser**
Basic block arguments are just added as operands to the argument list. The get() interface still needs some testing and the printing is not yet very intuitive.

### CODE REVIEW

File: src/xdsl/dialects/std.py [hunk 1]
```diff
@@ -41,3 +43,25 @@ class Return(Operation):
     @staticmethod
     def get(*ops: Union[Operation, SSAValue]) -> Return:
         return Return.build(operands=[[op for op in ops]])
+
+
+@irdl_op_definition
+class Branch(Operation):
+    name: str = "std.br"
```
Comment (webmiche) [inline src/xdsl/dialects/std.py:50]: I am uncertain how close we want to be with MLIR here, but AFAIK, MLIR has the `cf` dialect now for control flow instructions. Maybe add another dialect to this? Or don't and make sure to handle it in the MLIRConverter properly.

File: src/xdsl/dialects/std.py [hunk 2]
```diff
@@ -41,3 +43,25 @@ class Return(Operation):
...
+
+    @staticmethod
+    def get(*ops: Union[Operation, SSAValue], successor=Block) -> Branch:
+        return Branch.build(operands=[[op for op in ops]], successor=[Block])
+
+
+@irdl_op_definition
+class ConditionalBranch(Operation):
+    name: str = "std.cond_br"
+    arguments = VarOperandDef(AnyAttr())
```
Comment (webmiche) [inline src/xdsl/dialects/std.py:61]: I don't see how you know which target to pick. I would expect a flag operand (of type i1 I assume) to tell.

File: tests/std_test.py [hunk 3]
```diff
@@ -22,6 +22,20 @@
...
+  builtin.func() ["sym_name" = "br", "type" = !fun<[!i32], [!i32]>, "sym_visibility" = "private"]
+  {
+  ^2(%22: !i32):
+    std.br(%22: !i32)(^2)
+  }
+
+  builtin.func() ["sym_name" = "cond_br", "type" = !fun<[!i32], [!i32]>, "sym_visibility" = "private"]
+  {
+  ^3(%23 : !i32):
+    std.cond_br(%23 : !i32, %23 : !i32, %23 : !i32)(^3, ^4)
```
Comment (webmiche) [inline tests/std_test.py:35]: Same comment as above, how do you know where of the two possibilities to jump?

File: src/xdsl/dialects/std.py [hunk 4]
```diff
@@ -41,3 +43,25 @@ class Return(Operation):
...
         return Return.build(operands=[[op for op in ops]])
+
+
+@irdl_op_definition
+class Branch(Operation):
+    name: str = "std.br"
+    arguments = VarOperandDef(AnyAttr())
+
+    @staticmethod
+    def get(*ops: Union[Operation, SSAValue], successor=Block) -> Branch:
```
Comment (math-fehr) [inline src/xdsl/dialects/std.py:54]: Does this works as expected? Maybe we should move successor before ops?

**Review by webmiche — COMMENTED** [review]
LGTM generally, once again the discussion on how closely we should mirror the MLIR dialects comes up for me.

File: src/xdsl/dialects/std.py [hunk 5]
```diff
@@ -41,3 +43,25 @@ class Return(Operation):
...
+        return Branch.build(operands=[[op for op in ops]], successor=[Block])
+
+
+@irdl_op_definition
+class ConditionalBranch(Operation):
+    name: str = "std.cond_br"
+    arguments = VarOperandDef(AnyAttr())
+
+    @staticmethod
+    def get(*ops: Union[Operation, SSAValue],
```
Comment (math-fehr) [inline src/xdsl/dialects/std.py:64]: I think the best here would be to have `get(succ_true, [succ_true_args], succ_false, [succ_false_args])

**Review by math-fehr — APPROVED** [review]
Looks good to me besides the comments. I think we need support for custom printing/parsing in xDSL to make this print well.

File: src/xdsl/dialects/std.py (hunk 2 above)
Comment (math-fehr) [inline src/xdsl/dialects/std.py:61]: You should use two `VarOperandDef` here, to separate both block arguments.

File: src/xdsl/dialects/std.py (hunk 1 above)
Comment (tobiasgrosser) [reply src/xdsl/dialects/std.py:50]: I moved them to the cf dialect.

File: src/xdsl/dialects/std.py (hunk 2 above)
Comment (tobiasgrosser) [reply src/xdsl/dialects/std.py:61]: I added an i1 cond operand.

File: tests/std_test.py (hunk 3 above)
Comment (tobiasgrosser) [reply tests/std_test.py:35]: Changed.

File: src/xdsl/dialects/std.py (hunk 2 above)
Comment (tobiasgrosser) [reply src/xdsl/dialects/std.py:61]: Changed.

File: src/xdsl/dialects/std.py (hunk 5 above)
Comment (tobiasgrosser) [reply src/xdsl/dialects/std.py:64]: Changed.

File: src/xdsl/dialects/std.py (hunk 4 above)
Comment (tobiasgrosser) [reply src/xdsl/dialects/std.py]: I added some tests and it should now work as expected.

**Comment by tobiasgrosser** [comment]
OK, I did another pass on this PR. Did this address the open comments?

**Review by webmiche — APPROVED** [review]
LGTM

**Pull request merged**